### PR TITLE
LibGit2: tests & improvements

### PR DESCRIPTION
**Pull request by wildart**
PR provides following:
- more tests for `LibGit2` module (separated online part)
- `AbstractPayload` type for various callback payloads
- Credential callback can handle payload parameter (i.e. credentials)
- Payload support to high level functions: `clone`, `fetch`, `push`
- Some additional functions for remote references which are required for tests
- Function which generates for github credentials from git global config and 2FA token (if available)

### CODE REVIEW

File: test/libgit2.jl [hunk 1]
```diff
@@ -1,9 +1,12 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
-
```
Comment (tkelman) [inline test/libgit2.jl]: this first whitespace is considered significant and will get added back in by the license-headers script

File: test/libgit2.jl [hunk 2]
```diff
@@ -65,47 +36,115 @@ function temp_dir(fn::Function, remove_tmp_dir::Bool=true)
...
+        end
 
-    @test isdir(path)
-    @test isdir(joinpath(path, ".git"))
+        try
+            @testset "SSH protocol is not supported" begin
+                repo_path = joinpath(dir, "Example3")
+                @test_throws LibGit2.Error.GitError LibGit2.clone(ssh_prefix*repo_url, repo_path)
+            end
+        catch ex
```
Comment (tkelman) [inline test/libgit2.jl]: add a forced failure at the end of the try if it should always fail

**Comment by tkelman** [comment]
While you're refactoring the tests, can you move any that require internet access either to pkg or a separate libgit2-online test file? we cant run those by default but can add them to travis and appveyor

File: test/libgit2.jl [hunk 3]
```diff
@@ -41,206 +70,290 @@ begin
...
-    tmpdir = joinpath(tempdir(),randstring())
-    @test !isdir(tmpdir)
-    try
-        mkdir(tmpdir)
-        @test isdir(tmpdir)
-        fn(tmpdir)
-    finally
-        remove_tmp_dir && rm(tmpdir, recursive=true)
+    # test parameters
+    repo_url = "github.com/JuliaLang/Example.jl"
```
Comment (tkelman) [inline test/libgit2.jl]: cloning this requires net access, right? add libgit2-online to the travis and appveyor yml's

File: test/choosetests.jl [hunk 4]
```diff
@@ -32,7 +32,7 @@ function choosetests(choices = [])
         "nullable", "meta", "profile", "libgit2", "docs", "markdown",
         "base64", "serialize", "functors", "misc",
         "enums", "cmdlineargs", "i18n", "workspace", "libdl", "int",
-        "intset", "floatfuncs", "compile"
+        "intset", "floatfuncs", "compile", "libgit2-online"
```
Comment (tkelman) [inline test/choosetests.jl]: not here, note how pkg isn't in this list either

File: test/libgit2.jl (hunk 3 above)
Comment (wildart) [inline test/libgit2.jl]: I do not clone it here, it's used to setup remote reference only.

File: test/choosetests.jl (hunk 4 above)
Comment (wildart) [inline test/choosetests.jl]: where? in .travis.yml and appveyor.yml?

Comment (tkelman) [inline test/choosetests.jl]: yeah, where they currently do `runtests.jl pkg` separately, make that `runtests.jl pkg libgit2-online`

**Comment by wildart** [comment]
Ready for review.
cc @tkelman @jakebolewski @kshyatt

File: base/libgit2.jl [hunk 5]
```diff
@@ -232,13 +234,15 @@ function branch!(repo::GitRepo, branch_name::AbstractString,
             end
         end
 
-        # checout selected branch
-        with(peel(GitTree, branch_ref)) do btree
-            checkout_tree(repo, btree)
-        end
+        if set_head
+            # checout selected branch
```
Comment (tkelman) [inline base/libgit2.jl]: checkout

File: base/libgit2.jl [hunk 6]
```diff
@@ -289,17 +293,20 @@ function checkout!(repo::GitRepo, commit::AbstractString = "";
...
 
 """ git clone [-b <branch>] [--bare] <url> <dir> """
-function clone(repo_url::AbstractString, repo_path::AbstractString;
+function clone{P<:AbstractPayload}(repo_url::AbstractString, repo_path::AbstractString;
                branch::AbstractString="",
                isbare::Bool = false,
-               remote_cb::Ptr{Void} = C_NULL)
+               remote_cb::Ptr{Void} = C_NULL,
+               payload::Nullable{P}=Nullable{AbstractPayload}())
     # setup colne options
```
Comment (tkelman) [inline base/libgit2.jl]: clone

File: base/libgit2/callbacks.jl [hunk 7]
```diff
@@ -29,20 +22,45 @@ function mirror_callback(remote::Ptr{Ptr{Void}}, repo_ptr::Ptr{Void},
     return Cint(0)
 end
 
+"""Credentials callback function
+
+Function provides different credential acquisition functionality w.r.t. to a connection protocol.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: the t in w.r.t. already stands for "to"

File: base/libgit2/callbacks.jl [hunk 8]
```diff
@@ -29,20 +22,45 @@ function mirror_callback(remote::Ptr{Ptr{Void}}, repo_ptr::Ptr{Void},
...
+If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayload` object.
+
+For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields: `user` & `pass` they used to create authentication credentials.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: they are used to create

**Comment by wildart** [comment]
Squashed it. I need to merge it to continue work on #13387 and [PkgDev](https://github.com/JuliaLang/PkgDev.jl)

File: base/libgit2/callbacks.jl [hunk 9]
```diff
@@ -29,20 +22,45 @@ function mirror_callback(remote::Ptr{Ptr{Void}}, repo_ptr::Ptr{Void},
...
 
     if isset(allowed_types, Cuint(Consts.CREDTYPE_USERPASS_PLAINTEXT))
-        # use keyboard-interactive prompt
-        username = prompt("Username for '$url'")
-        pass     = prompt("Password for '$url'", password=true)
+        username = password = ""
+        if payload_ptr != C_NULL
+            payload = unsafe_pointer_to_objref(payload_ptr)
+            if isa(payload, AbstractPayload)
+                username = isdefined(payload, :user) ? getfield(payload, :user) : ""
```
Comment (jakebolewski) [inline base/libgit2/callbacks.jl]: I feel that checking for specific field names for a given type should be wrapped in a generic function: `user(::Payload)`, `password(::Payload)`, `isused(::Payload)`, etc.

File: base/libgit2.jl [hunk 10]
```diff
@@ -289,17 +293,20 @@ function checkout!(repo::GitRepo, commit::AbstractString = "";
...
+               payload::Nullable{P}=Nullable{AbstractPayload}())
+    # setup clone options
+    fetch_opts=FetchOptions(callbacks = RemoteCallbacks(credentials_cb(), payload))
     clone_opts = CloneOptions(
-                    bare = Int32(isbare),
-                    checkout_branch = isempty(branch) ? Cstring_NULL :
-                                      convert(Cstring, pointer(branch)),
-                    remote_cb = remote_cb
-                )
+                bare = Int32(isbare),
```
Comment (jakebolewski) [inline base/libgit2.jl]: Should this be a `Cint`?

File: base/libgit2/consts.jl [hunk 11]
```diff
@@ -263,4 +263,12 @@ module Consts
                         CREDTYPE_SSH_INTERACTIVE    = Cuint(1 << 4),
                         CREDTYPE_USERNAME           = Cuint(1 << 5),
                         CREDTYPE_SSH_MEMORY         = Cuint(1 << 6))
+
+    @enum(GIT_CONFIG, CONFIG_LEVEL_DEFAULT = 0,
```
Comment (jakebolewski) [inline base/libgit2/consts.jl]: Could these inline comments be moved into docstrings?

File: base/libgit2.jl (hunk 10 above)
Comment (wildart) [inline base/libgit2.jl]: Yes, it should.

File: base/libgit2/callbacks.jl (hunk 9 above)
Comment (wildart) [inline base/libgit2/callbacks.jl]: Good idea.

File: base/libgit2/consts.jl (hunk 11 above)
Comment (wildart) [inline base/libgit2/consts.jl]: I'll do it, but I do not feel like putting every enum value into docs. Enum values usually are self-explanatory.

Comment (jakebolewski) [inline base/libgit2/consts.jl]: I wasn't really suggesting that.  If you are going to write inline code documentation for enum values, why not get docstrings as well for free?

Comment (wildart) [inline base/libgit2/consts.jl]: how do you do inline docs for enums?

Comment (ScottPJones) [inline base/libgit2/consts.jl]: Just put them in a string *before* the field to be documented.

Comment (wildart) [inline base/libgit2/consts.jl]: I see for enum type, but not its values.

**Comment by tkelman** [comment]
`@testset` is kind of noisy, we should shut off the summary results display when everything succeeds

**Pull request merged**